DEVDOCS.md

```diff
@@ -0,0 +1,92 @@
+# Enzyme-JAX Developer Notes
+
+## Building the Project
+
+### Quick Build
+```bash
+bazel build --repo_env=CC=clang-18 --color=yes -c dbg :enzymexlamlir-opt
+```
+
+### Build Artifacts
+- **Main tool**: `enzymexlamlir-opt` (bazel target: `:enzymexlamlir-opt`)
+  - This is the MLIR optimization tool driver for Enzyme-XLA
+  - Analogous to `mlir-opt`, drives compiler passes and transformations
+  - Located in: `src/enzyme_ad/jax/enzymexlamlir-opt.cpp`
+
+- **Python wheel**: `bazel build :wheel`
+
+### Generate LSP Support
+```bash
+bazel run :refresh_compile_commands
+```
+
+## Project Structure
+
+### Core Components
+
+#### 1. **Dialects** (`src/enzyme_ad/jax/Dialect/`)
+MLIR dialects define custom operations and types for Enzyme-JAX.
+
+- **EnzymeXLAOps.td** - Dialect operation definitions
+  - GPU operations: `kernel_call`, `memcpy`, `gpu_wrapper`, `gpu_block`, `gpu_thread`
+  - JIT/XLA operations: `jit_call`, `xla_wrapper`
+  - Linear algebra (BLAS/LAPACK): `symm`, `syrk`, `trmm`, `lu`, `getrf`, `gesvd`, etc.
+  - Special functions: Bessel functions, GELU, ReLU
+  - Utility operations: `memref2pointer`, `pointer2memref`, `subindex`
+
+- **EnzymeXLAAttrs.td** - Custom attribute definitions (LAPACK enums, etc.)
+
+#### 2. **Passes** (`src/enzyme_ad/jax/Passes/`)
+MLIR passes implement transformations and optimizations.
+
+- Tablegen definitions in `src/enzyme_ad/jax/Passes/Passes.td`
+- **EnzymeHLOOpt.cpp** - Core optimization patterns for StableHLO and EnzymeXLA operations
+
+#### 3. **Transform Operations** (`src/enzyme_ad/jax/TransformOps/`)
+In order to have more granular control over which pattern is applied, patterns are also registered as transform operations.
+For example:
+```
+def AndPadPad : EnzymeHLOPatternOp<
+    "and_pad_pad"> {
+  let patterns = ["AndPadPad"];
+}
+```
+Exposes the `AndPadPad` pattern (defined in `EnzymeHLOOpt.cpp`) to `enzymexlamlir-opt`, so it can be used as:
+```
+enzymexlamlir-opt --enzyme-hlo-generate-td="patterns=and_pad_pad" --transform-interpreter --enzyme-hlo-remove-transform -allow-unregistered-dialect input.mlir
+```
+
+## Common Development Tasks
+
+### Adding a New Optimization Pattern
+
+1. Define the pattern class in `src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp`
+2. Inherit from `mlir::OpRewritePattern<OpType>`
+3. Implement `matchAndRewrite()` method
+4. Register in `EnzymeHLOOptPass::runOnOperation()`
+5. Register as Transform operation in `TransformOps.td`
```

Comment:
We can also say to add to primitives.py which I often forget

Reply:
Thanks! Feel free to just push, otherwise I can do it later. I didn't even know about primitives.py 🙈